From 21be8e8685c60371dc83a97684f038e462f9da91 Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Mon, 3 Nov 2014 21:42:46 -0500
Subject: [powerpc] pseries: Fix endian issues in onlining cpu threads

Message-id: <20141103214246.10436.20776.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 99374
O-Subject: [PATCH RHEL7.1 BZ1159579 1/3] pseries: Fix endian issues in onlining cpu threads
Bugzilla: 1159579
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1159579

Description:
------------
The ibm,ppc-interrupt-server#s property is in big endian format.
These values need to be converted when used by little endian
architectures.

Signed-off-by: Thomas Falcon <tlfalcon@linux.vnet.ibm.com>
Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>

RHEL Version Found:
-------------------
7.0

Upstream Status:
----------------
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=822e71224e07f07a07c385be869fe416ce436430

----------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/pseries/dlpar.c b/arch/powerpc/platforms/pseries/dlpar.c
index 2d0b4d6..3eb3781 100644
--- a/arch/powerpc/platforms/pseries/dlpar.c
+++ b/arch/powerpc/platforms/pseries/dlpar.c
@@ -363,7 +363,8 @@ static int dlpar_online_cpu(struct device_node *dn)
  int rc = 0;
  unsigned int cpu;
  int len, nthreads, i;
- const u32 *intserv;
+ const __be32 *intserv;
+ u32 thread;
 
  intserv = of_get_property(dn, "ibm,ppc-interrupt-server#s", &len);
  if (!intserv)
@@ -373,8 +374,9 @@ static int dlpar_online_cpu(struct device_node *dn)
 
  cpu_maps_update_begin();
  for (i = 0; i < nthreads; i++) {
+  thread = be32_to_cpu(intserv[i]);
   for_each_present_cpu(cpu) {
-   if (get_hard_smp_processor_id(cpu) != intserv[i])
+   if (get_hard_smp_processor_id(cpu) != thread)
     continue;
    BUG_ON(get_cpu_current_state(cpu)
      != CPU_STATE_OFFLINE);
@@ -388,7 +390,7 @@ static int dlpar_online_cpu(struct device_node *dn)
   }
   if (cpu == num_possible_cpus())
    printk(KERN_WARNING "Could not find cpu to online "
-          "with physical id 0x%x\n", intserv[i]);
+          "with physical id 0x%x\n", thread);
  }
  cpu_maps_update_done();
 
-- 
1.7.1